### DISCUSSION TITLE
How to add group headers

### DISCUSSION THREAD

**Discussion by olivermicke** [TanStack/table#3320] [Q&A]
Hey,
first of all thank you for creating this awesome library.

How would you recommend to implement group headers [Example image](https://i.stack.imgur.com/3HPH9.png)

I was unable to use the built-in grouping as I don't want to aggregate any data, but rather add more (group header) rows.

**Comment by DPCamino** [comment]
Were you able to achieve this result? I too was trying to achieve a [similar look](https://miro.medium.com/max/2000/1*S6an7XQsdMyUBUt9nH2L-g.png).

**Reply by olivermicke** [reply]
Unfortunately not. Best of luck to you

**Reply by aprilchang** [reply]
Did either of you figure this out?  We are trying to do the same thing and would appreciate tips if you have any.  Thanks.

**Reply by DPCamino** [reply]
Yup! 

Basically, create a new `tbody` for every group, which according to documentation is an appropriate `table` structure:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/tbody#multiple_bodies

Map through a unique group array (that you create from your data) and render each `tbody` section with its own `th` and `tr` elements. Here is example:

```tsx

            uniqueGroupsArray.map(group =>
              <tbody {...getTableBodyProps()}>
                  <tr>
                    <th colSpan={visibleColumns.length}>{group}</th>
                  </tr>
                {rows.filter(x => x.original.DataProperty == group).map(row => {
                  prepareRow(row);
                  return (
                    <tr {...row.getRowProps()}>
                      {row.cells.map(cell => {
                        return (
                          <td {...cell.getCellProps()}>{cell.render('Cell')}</td>
                        );
                      })}
                    </tr>
                  );
                })}
              </tbody>
            )

```

You map through an array of group header names that you create ( `uniqueGroupsArray` ), rendering the name of each group header in its own `th`, then filter and render table data that matches that particular header group, e.g., `DataProperty`.

**Reply by aprilchang** [reply]
Thank you!

**Reply by verylovestars** [reply]
Is this a current solution?

**Reply by jacksonkasi1** [reply]
> Is this a current solution?

Repo: https://github.com/jacksonkasi1/FlexiSheet

Demo: https://flexisheet.vercel.app/

Blog: https://dev.to/jacksonkasi/building-a-google-sheets-like-table-component-with-tanstack-table-zod-and-shadcnui-1ped

**Comment by jacksonkasi1** [comment]
Repo: https://github.com/jacksonkasi1/FlexiSheet

Demo: https://flexisheet.vercel.app/

Blog: https://dev.to/jacksonkasi/building-a-google-sheets-like-table-component-with-tanstack-table-zod-and-shadcnui-1ped
